extensions/amp-viewer-integration/0.1/findtext.js

```diff
@@ -0,0 +1,451 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {computedStyle} from '../../../src/style';
+
+/**
+ * Simple implementation of CircularBuffer.
+ * Exported for test only.
+ */
+export class CircularBuffer {
+  constructor(max) {
+    this.max_ = max;
+    this.buff_ = [];
+    this.next_ = 0;
+  }
+
+  add(item) {
+    this.buff_[this.next_] = item;
+    this.next_ = (this.next_ + 1) % this.max_;
+  }
+
+  /**
+   * @param {number} index The index of an element to get.
+   */
+  get(index) {
+    if (this.buff_.length >= this.max_) {
+      index = (this.next_ + index) % this.max_;
+    }
+    return this.buff_[index];
+  }
+}
+
+/**
+ * TextPos is a pointer to a character in a Text node.
+ * Exported for test only.
+ */
+export class TextPos {
+  /**
+   * @param {!Text} node
+   * @param {number} offset
+   */
+  constructor(node, offset) {
+    this.node = node;
+    this.offset = offset;
+  }
+
+  /**
+   * @return {number}
+   */
+  get char() {
+    return this.node.wholeText[this.offset];
+  }
+}
+
+/**
+ * TextRange represents a text range.
+ * Exported for test only.
+ */
```

Comment:
Done. Do you think this is better than class?
https://github.com/ampproject/amphtml/pull/15337/commits/14f618ea8ee57717b5b264306f32660f9f0ca079

Reply:
Much better!

AMP style for `typedef` variables is to suffix with `Def`, i.e. `TextRangeDef`. Otherwise the linter will complain about the variable not being in use.

Reply:
Renamed TextRange/TextPos to TextRangeDef/TextPosDef.